File: Sources/VIPSImage+Transform.m
//
//  VIPSImage+Transform.m
//  VIPSKit
//
//  Image transformation methods (crop, rotate, flip)
//

#import "VIPSImage+Private.h"

@implementation VIPSImage (Transform)

- (VIPSImage *)cropWithX:(NSInteger)x
                       y:(NSInteger)y
                   width:(NSInteger)width
                  height:(NSInteger)height
                   error:(NSError **)error {
    VipsImage *out = NULL;

    if (vips_crop(self.image, &out, (int)x, (int)y, (int)width, (int)height, NULL) != 0) {
        if (error) {
            *error = [self.class errorFromVips];
        }
        return nil;
    }

    VIPSImage *result = [[VIPSImage alloc] init];
    result.image = out;
    return result;
}

- (VIPSImage *)rotateByDegrees:(NSInteger)degrees error:(NSError **)error {
    VipsImage *out = NULL;
    VipsAngle angle;

    switch (degrees % 360) {
        case 90:
        case -270:
            angle = VIPS_ANGLE_D90;
            break;
        case 180:
        case -180:
            angle = VIPS_ANGLE_D180;
            break;
        case 270:
        case -90:
            angle = VIPS_ANGLE_D270;
            break;
        default:
            angle = VIPS_ANGLE_D0;
            break;
    }

    if (vips_rot(self.image, &out, angle, NULL) != 0) {
        if (error) {
            *error = [self.class errorFromVips];
        }
        return nil;
    }

    VIPSImage *result = [[VIPSImage alloc] init];
    result.image = out;
    return result;
}

- (VIPSImage *)flipHorizontalWithError:(NSError **)error {
    VipsImage *out = NULL;

    if (vips_flip(self.image, &out, VIPS_DIRECTION_HORIZONTAL, NULL) != 0) {
        if (error) {
            *error = [self.class errorFromVips];
        }
        return nil;
    }

    VIPSImage *result = [[VIPSImage alloc] init];
    result.image = out;
    return result;
}

- (VIPSImage *)flipVerticalWithError:(NSError **)error {
    VipsImage *out = NULL;

    if (vips_flip(self.image, &out, VIPS_DIRECTION_VERTICAL, NULL) != 0) {
        if (error) {
            *error = [self.class errorFromVips];
        }
        return nil;
    }

    VIPSImage *result = [[VIPSImage alloc] init];
    result.image = out;
    return result;
}

- (VIPSImage *)autoRotateWithError:(NSError **)error {
    VipsImage *out = NULL;

    if (vips_autorot(self.image, &out, NULL) != 0) {
        if (error) {
            *error = [self.class errorFromVips];
        }
        return nil;
    }

    VIPSImage *result = [[VIPSImage alloc] init];
    result.image = out;
    return result;
}

@end

File: VIPSImage+Transform.swift
import Foundation
@_implementationOnly import vips
@_implementationOnly import CVIPS

extension VIPSImage {

    /// Crop a region from the image.
    public func crop(x: Int, y: Int, width: Int, height: Int) throws -> VIPSImage {
        var out: UnsafeMutablePointer<VipsImage>?
        guard cvips_crop(pointer, &out, Int32(x), Int32(y), Int32(width), Int32(height)) == 0,
              let out else { throw VIPSError.fromVips() }
        return VIPSImage(pointer: out)
    }

    /// Rotate by 90, 180, or 270 degrees.
    public func rotate(degrees: Int) throws -> VIPSImage {
        let angle: VipsAngle
        switch degrees % 360 {
        case 90, -270:   angle = VIPS_ANGLE_D90
        case 180, -180:  angle = VIPS_ANGLE_D180
        case 270, -90:   angle = VIPS_ANGLE_D270
        default:         angle = VIPS_ANGLE_D0
        }

        var out: UnsafeMutablePointer<VipsImage>?
        guard cvips_rot(pointer, &out, angle) == 0, let out else {
            throw VIPSError.fromVips()
        }
        return VIPSImage(pointer: out)
    }

    /// Mirror horizontally.
    public func flipHorizontal() throws -> VIPSImage {
        var out: UnsafeMutablePointer<VipsImage>?
        guard cvips_flip(pointer, &out, VIPS_DIRECTION_HORIZONTAL) == 0, let out else {
            throw VIPSError.fromVips()
        }
        return VIPSImage(pointer: out)
    }

    /// Mirror vertically.
    public func flipVertical() throws -> VIPSImage {
        var out: UnsafeMutablePointer<VipsImage>?
        guard cvips_flip(pointer, &out, VIPS_DIRECTION_VERTICAL) == 0, let out else {
            throw VIPSError.fromVips()
        }
        return VIPSImage(pointer: out)
    }

    /// Auto-rotate based on EXIF orientation.
    public func autoRotate() throws -> VIPSImage {
        var out: UnsafeMutablePointer<VipsImage>?
        guard cvips_autorot(pointer, &out) == 0, let out else {
            throw VIPSError.fromVips()
        }
        return VIPSImage(pointer: out)
    }

    /// Content-aware smart crop.
    public func smartCrop(toWidth width: Int, height: Int, interesting: Interesting = .attention) throws -> VIPSImage {
        let vipsInteresting: VipsInteresting
        switch interesting {
        case .none:      vipsInteresting = VIPS_INTERESTING_NONE
        case .centre:    vipsInteresting = VIPS_INTERESTING_CENTRE
        case .entropy:   vipsInteresting = VIPS_INTERESTING_ENTROPY
        case .attention: vipsInteresting = VIPS_INTERESTING_ATTENTION
        case .low:       vipsInteresting = VIPS_INTERESTING_LOW
        case .high:      vipsInteresting = VIPS_INTERESTING_HIGH
        }

        var out: UnsafeMutablePointer<VipsImage>?
        guard cvips_smartcrop(pointer, &out, Int32(width), Int32(height), vipsInteresting) == 0,
              let out else { throw VIPSError.fromVips() }
        return VIPSImage(pointer: out)
    }
}
